src/explosion.cpp

```diff
@@ -435,93 +435,115 @@ std::vector<tripoint> game::shrapnel( const tripoint &src, int power,
+            if( cloud.density <= MIN_FRAGMENT_DENSITY ||
+                cloud.velocity <= MIN_EFFECTIVE_VELOCITY ) {
+                continue;
+            }
+            distrib.emplace_back( x, y, src.z );
+            tripoint target( x, y, src.z );
+            int damage = ballistic_damage( cloud.velocity, fragment_mass );
+            auto critter = critter_at( target );
+            if( damage > 0 && critter && !critter->is_dead_state() ) {
+                static std::default_random_engine eng(
+                    std::chrono::system_clock::now().time_since_epoch().count() );
+                std::poisson_distribution<> d( cloud.density );
+                int hits = d( eng );
+                dealt_projectile_attack frag;
+                frag.proj = proj;
+                frag.proj.speed = cloud.velocity;
+                frag.proj.impact = damage_instance::physical( 0, damage, 0, 0 );
+                // dealt_dam.total_damage() == 0 means armor block
+                // dealt_dam.total_damage() > 0 means took damage
+                // Need to diffentiate target among player, npc, and monster
+                // Do we even print monster damage?
+                int damage_taken = 0;
+                int damaging_hits = 0;
+                int non_damaging_hits = 0;
+                for( int i = 0; i < hits; ++i ) {
+                    frag.missed_by = rng_float( 0.05, 1.0 );
+                    critter->deal_projectile_attack( nullptr, frag, false );
+                    if( frag.dealt_dam.total_damage() > 0 ) {
+                        damaging_hits++;
+                        damage_taken += frag.dealt_dam.total_damage();
+                    } else {
+                        non_damaging_hits++;
+                    }
+                    add_msg( m_debug, "Shrapnel hit %s at %d m/s at a distance of %d",
+                             critter->disp_name().c_str(),
+                             frag.proj.speed, rl_dist( src, target ) );
+                    add_msg( m_debug, "Shrapnel dealt %d damage", frag.dealt_dam.total_damage() );
+                    if( critter->is_dead_state() ) {
+                        break;
                     }
                 }
-                if( m.impassable( target ) ) {
-                    if( optional_vpart_position vp = m.veh_at( target ) ) {
-                        vp->vehicle().damage( vp->part_index(), damage );
+                int total_hits = damaging_hits + non_damaging_hits;
+                if( total_hits > 0 && g->u.sees( *critter ) ) {
+                    // Building a phrase to summarize the fragment effects.
+                    // Target, Number of impacts, total amount of damage, proportion of deflected fragments.
+                    std::map<int, std::string> impact_count_descriptions = {
+                        { 1, _( "a" ) }, { 2, _( "several" ) }, { 5, _( "many" ) },
+                        { 20, _( "a large number" ) }, { 100, _( "a huge number" ) }
```

Comment:
As this will be used in, e.g., "[You are/NPC is/The Monster is] hit by %s bomb fragments", I think you want "a [large/huge] number ***of***".